apps/src/dance/LottieDancerRenderer.ts

```diff
@@ -0,0 +1,673 @@
+              headInfo.h
+            );
+            this._insertHeadImageLayer(
+              headComp,
+              insertIndex,
+              assetId,
+              headInfo.w,
+              headInfo.h,
+              headKs
+            );
+          }
+        }
+      }
+      // Memoize transformed Lottie JSON per move (in-memory cache) so subsequent setSource calls skip recolor/head work.
+      // This is useful if the same dance move is used later in a song, or if there are multiple generated dancers using the same move.
+      this.cachedAnimationData[move] = animData;
+    }
+
+    // Lottie instance bound to our canvas 2D context
+    await this._prepareLottie(animData);
+
+    this.animationData = animData;
+    this.totalFrames = Math.max(
+      0,
+      Math.round((animData.op || 0) - (animData.ip || 0))
+    );
+
+    // Optionally capture comp size if top-level provides it.
+    if (typeof animData.w === 'number' && typeof animData.h === 'number') {
+      this.compW = animData.w;
+      this.compH = animData.h;
+    }
+  }
+
+  getDurationFrames(): number | null {
+    return this.totalFrames;
+  }
+
+  getCompSize(): {w: number; h: number} | null {
+    return this.compW && this.compH ? {w: this.compW, h: this.compH} : null;
+  }
+
+  renderFrame(frameIndex: number): void {
+    if (!this.anim || !this.ctx || this.totalFrames === null) {
+      return;
+    }
+    const totalFrames = Math.max(1, this.totalFrames || 1);
+    const frame = Math.floor(
+      ((frameIndex % totalFrames) + totalFrames) % totalFrames
+    );
+
+    this.anim.goToAndStop(frame, true);
+  }
+
+  getTotalFrames(): number | null {
+    return this.totalFrames;
+  }
+
+  resize(): void {
+    if (!this.anim) return;
+    // ExternalDancerLayer may recreate the graphics; update renderer references.
+    this.anim.renderer.ctx = this.ctx;
+    if (typeof this.anim.resize === 'function') this.anim.resize();
+  }
+
+  dispose(): void {
+    this._destroyAnim();
+  }
+
+  // URL helpers
+  private _resolveAnimationUrl(danceMove: string): string {
+    // Example: .../dancers/input/DUCK/duck_<move>.json
+    return `${BASE_HOST}/dancers/input/${this.skeletonName.toUpperCase()}/${
+      this.skeletonName
+    }_${danceMove}.json`;
+  }
+
+  private _resolveMetadataUrl(): string {
+    // Example: .../dancer/<assetsPath>/<dancerName>-metadata.json
+    return `${BASE_HOST}/dancer/${this.assetsPath}/${this.dancerName}-metadata.json`;
+  }
+
+  private _resolveHeadUrl(): string {
+    // Example: .../dancer/<assetsPath>/<dancerName>.png
+    return `${BASE_HOST}/dancer/${this.assetsPath}/${this.dancerName}.png`;
+  }
+
+  // Lottie lifecycle
+  private async _prepareLottie(animationData: LottieJSON): Promise<void> {
+    this._destroyAnim();
+    const config: CanvasAnimConfig = {
+      renderer: 'canvas',
+      // Not used by the canvas renderer, but required by the typings.
+      //   container: document.createElement('div'),
```

Comment:
Required but commented out?

Reply:
Oh, thanks. I fixed this by casting the `loadAnimation` function to get around the typings. It's not actually required. Will remove.